things/service/src/main/java/org/eclipse/ditto/things/service/utils/ReadGrantCollector.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2025 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.things.service.utils;
+
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.ditto.base.model.auth.AuthorizationContext;
+import org.eclipse.ditto.base.model.auth.AuthorizationSubject;
+import org.eclipse.ditto.base.model.auth.DittoAuthorizationContextType;
+import org.eclipse.ditto.json.JsonFieldSelector;
+import org.eclipse.ditto.json.JsonObject;
+import org.eclipse.ditto.json.JsonPointer;
+import org.eclipse.ditto.json.JsonValue;
+import org.eclipse.ditto.policies.api.Permission;
+import org.eclipse.ditto.policies.enforcement.PolicyEnforcer;
+import org.eclipse.ditto.policies.model.Permissions;
+import org.eclipse.ditto.policies.model.PoliciesResourceType;
+import org.eclipse.ditto.policies.model.ResourceKey;
+import org.eclipse.ditto.policies.model.enforcers.Enforcer;
+import org.eclipse.ditto.things.model.Thing;
+
+/**
+ * Utility class for collecting read grants from a PolicyEnforcer.
+ * Determines which subjects can read which paths for a given set of fields.
+ */
+public final class ReadGrantCollector {
+
+    private ReadGrantCollector() {
+        // No instantiation
+    }
+
+    /**
+     * Collects read grants for the specified fields from the enforcer.
+     *
+     * @param fields the fields to collect grants for
+     * @param thing the Thing entity
+     * @param policyEnforcer the PolicyEnforcer to query
+     * @return ReadGrant mapping paths to sets of subject IDs
+     */
+    public static ReadGrant collect(
+            final JsonFieldSelector fields,
+            final Thing thing,
+            final PolicyEnforcer policyEnforcer
+    ) {
+        if (fields == null || thing == null || policyEnforcer == null) {
```

Comment:
All parameters are not marked with `@Nullable` - so I would expect this can never evaluate to `true`.
If some parameters can be null, they must be declared so ..